### PR TITLE
adds quantized implementation of hard sigmoid

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34607 adds quantized implementation of hard sigmoid**

Summary:

Adds quantized version of hardsigmoid activation.

Note: not implementing the _ and .out versions is
currently intended, because the implementation changes the scale and
zp and it's nice to not allow the user to specify scale
and zp.  Lmk if we should handle this differently.

Test Plan:

tests
benchmarks

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20480546](https://our.internmc.facebook.com/intern/diff/D20480546)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -306,6 +306,69 @@ void qsigmoid_kernel(const Tensor& qx, Tensor& qy) {
...
+  float scale = qx.q_scale();
+  auto scale_vec = Vec256<float>(scale);
+  auto zero_point_vec = Vec256<float>((float)zero_point);
+  auto scale_neg_zp_premul_vec = scale_vec * zero_point_vec.neg();
+
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "qhardsigmoid", [&]() {
+
+    // Naive implemenentation: uses dequantize/execute/quantize routine
+    // - Output scale is set to 1.0 / 2^(BIT_NUM)
+    // - For signed types output zero point is set to 0
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: I think this comment and the one below should be reversed. We set output zero_point to 0 for unsigned types, right?

**Comment by supriyar** [comment]
LGTM, do we have some perf numbers as well for this?
I wonder if it would be possible to get some benchmark numbers on mobile as well for the elementwise operators you've implemented so far. Vec256 wouldn't give any speedup on mobile and it would be nice to track what the perf is on mobile as well for future optimizations.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 2]
```diff
@@ -306,6 +306,69 @@ void qsigmoid_kernel(const Tensor& qx, Tensor& qy) {
...
+    fVec kThreeVec(3.0f);
+    fVec kSixVec(6.0f);
+
+    cpu_kernel_vec(
+      iter,
+      [&](scalar_t qx) -> scalar_t {
+        auto x = at::dequantize_val(scale, zero_point, qx);
+        x = x + 3.0f;
+        x = x > 0.0f ? x : 0.0f;
+        auto y = (x < 6.0f ? x : 6.0f) / 6.0f;
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Same comment as the other PR, can do this branch free

**Comment by vkuzo** [comment]
> LGTM, do we have some perf numbers as well for this?
> I wonder if it would be possible to get some benchmark numbers on mobile as well for the elementwise operators you've implemented so far. Vec256 wouldn't give any speedup on mobile and it would be nice to track what the perf is on mobile as well for future optimizations.

sure thing, here is the output of the current benchmark: https://our.internmc.facebook.com/intern/paste/P127415136/

I will follow up offline on the best way to benchmark all of these things on mobile!

**Review by supriyar — APPROVED** [review]
Thanks for the benchmarks! I usually also compare against float to see the speedup. 
But looks good overall

**Comment by mrshenli** [comment]
Looks like this PR breaks master:

```
Mar 18 00:56:58 ======================================================================
Mar 18 00:56:58 FAIL [0.974s]: test_qhardsigmoid (__main__.TestQuantizedOps)
Mar 18 00:56:58 ----------------------------------------------------------------------
Mar 18 00:56:58 Traceback (most recent call last):
Mar 18 00:56:58   File "test_quantized.py", line 244, in test_qhardsigmoid
Mar 18 00:56:58     elements=hu.floats(-1e3, 1e3, allow_nan=False, allow_infinity=False),
Mar 18 00:56:58   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 1116, in wrapped_test
Mar 18 00:56:58     raise the_error_hypothesis_found
Mar 18 00:56:58   File "test_quantized.py", line 271, in test_qhardsigmoid
Mar 18 00:56:58     message="Hardsigmoid failed: {} vs. {}".format(qY, qY_hat))
Mar 18 00:56:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 880, in assertEqual
Mar 18 00:56:58     exact_dtype=exact_dtype)
Mar 18 00:56:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 882, in assertEqual
Mar 18 00:56:58     assertTensorsEqual(x, y)
Mar 18 00:56:58   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 848, in assertTensorsEqual
Mar 18 00:56:58     self.assertLessEqual(max_err, prec, message)
Mar 18 00:56:58 AssertionError: tensor(1, dtype=torch.int32) not less than or equal to 1e-05 : Hardsigmoid failed: tensor([0.5000], size=(1,), dtype=torch.qint32,
Mar 18 00:56:58        quantization_scheme=torch.per_tensor_affine, scale=2.3283064365386963e-10,
Mar 18 00:56:58        zero_point=0) vs. tensor([-0.5000], size=(1,), dtype=torch.qint32,
Mar 18 00:56:58        quantization_scheme=torch.per_tensor_affine, scale=2.3283064365386963e-10,
Mar 18 00:56:58        zero_point=0)
```

e.g.:

https://app.circleci.com/pipelines/github/pytorch/pytorch/141312/workflows/9036f70f-a41d-4968-aedc-4cab6b4307f0/jobs/4849783
https://app.circleci.com/pipelines/github/pytorch/pytorch/141312/workflows/9036f70f-a41d-4968-aedc-4cab6b4307f0/jobs/4849602

**Comment by mrshenli** [comment]
@vkuzo I am reverting this PR to fix test signals on master. Please resubmit with a fix. Thanks!

**Comment by vkuzo** [comment]
> @vkuzo I am reverting this PR to fix test signals on master. Please resubmit with a fix. Thanks!

sorry about that, will do!

**Pull request closed without merging**